eclipse-theia/theia
# update to Electron 15.1.2 and support Node.js 14

**What it does**

We currently have a hard dependency on `electron`, meaning that once
packaged we have a redundant `electron` installation part of the
application.

A workaround has been to configure the packagers to ignore that runtime
dependency on `electron`, but this is a workaround that has to be done
for each app.

In order to make it as easy as possible for adopters to consume the
framework, the goal is to encapsulate Electron-specific dependencies
into `@theia/electron` and have `electron` itself added to application's
`package.json` under `devDependencies`.

As part of updating the version of `electron`, this commit also fixes
the dependency problems in the framework:

- Update `electron` from `9.4.4` to `^15.3.5`.
- Move the `ffmepg` related C code into its own `@theia/ffmpeg` package.
- Move the `ffmpeg` replace and check scripts from `@theia/electron` to
  `@theia/cli` as commands: `ffmpeg:replace` and `ffmpeg:check`.
- Run the FFMPEG library replacing for Electron when doing `theia
  build` on the `electron` target.
- Add `theia-re-exports` utility to generate re-exports and generate
  documentation (based on `mustache` templates).
- Break `@theia/electron` to re-export Electron-specific dependencies as
  `@theia/electron/shared/...`.
- Modify `@theia/core` re-export scripts to re-export everything from
  `@theia/electron/shared/...` as `@theia/core/electron-shared/...`.
- Update the `@theia/eslint-plugin` rule that helps with re-using shared
  exports within the code base.
- Modify the generators to automatically add `electron` to
  `devDependencies` for Electron applications.
- Add utilities to detect or set `electron-rebuild`'s ABI when
  rebuilding modules for Electron as it might get lost with versions
  that are too recent.
- Re-export `ajv` from `@theia/core`.
- Update the `package.json` scripts to silence `yarn`, otherwise it will
  stack error messages when nested `yarn run` calls are executed.

#### How to test

- Everything should work like before.
- Delete the `electron` entry from `examples/electron/package.json#devDependencies`.
- Running `yarn build` in the Electron example application should add it back.
- Delete the `@theia/electron` entry from `examples/electron/package.json#dependencies`.
- Running `yarn build` in the Electron example application should fail because of the missing dependency.

#### Review checklist

- [x] [electron cq](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23785)
- [x] [ffmpeg cq](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23786)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
